Add support for simple export protocol

Wolf480pl opened the pull request:
as specified in:  https://git-annex.branchable.com/design/external_special_remote_protocol/export_and_import_appendix/

I believe if merged, this would close #30 and #35.

I extracted some code into the `do_checkpresent` and `do_remove` functions to avoid copy-pasting - should that go in a separate commit, or is it ok as is?

I've tested the new functionality locally with an FTP remote, but haven't written any automated tests for it. I'll try to add some if necessary.

Wolf480pl committed:
Add support for simple export protocol

as specified in:  https://git-annex.branchable.com/design/external_special_remote_protocol/export_and_import_appendix/

DanielDent commented:
@Wolf480pl - thank you so much! I know a lot of people have wanted this functionality, and I really appreciate your work.

Yes, if if it's easy for you, I'd prefer to see the `do_checkpresent` / `do_remove` refactoring done in a different commit - that makes it easier to review/understand the changes/commit history.

While I'm somewhat inclined towards simply merging your PR, I also want to be respectful that I've had contributors improving our test coverage, and I'd like to try to maintain that standard when we add features - after all, users are entrusting us with their data. I'd really appreciate seeing some tests for this feature, that would make me feel good about merging this.

Wolf480pl commented:
> thank you so much! I know a lot of people have wanted this functionality

I need it too, that's why I implemented it :wink:

I split off the refactoring into a separate commit. Let me know if there's anything else I should change.

I totally understand wanting to have this covered with tests. I'll try to add some in the next couple days, and also document this feature in the readme. Now that I think of it, there should be a warning to only use it with rclone backends that support atomic `moveto`, I'll add that too.

yarikoptic commented:
> I split off the refactoring into a separate commit. Let me know if there's anything else I should change.

made it into #63 for individual review/merge.

yarikoptic commented:
I merged #63, please rebase (e.g. via `git pull --rebase origin master` if this repo is your `origin`, otherwise change to some other name)

Wolf480pl commented:
I hit a problem:

in `TRANSFEREXPORT STORE` I use the git-annex key as temporary name for the file, before `rclone move`-ing it to the final destination. That's no good - the key isn't unique (eg. there could be 2 identical files with different names, or there could be 2 clones of the repo trying to export at the same time).

I could add eg. hash of the export path to the temporary name to make it unique within one invocation of `git annex export` but that still can collide if two such commands are running in parallel, possibly on different clones of the repo.

I was thinking adding annex UUID of the repo that's doing the export would be unique enough, but I don't know if `GETGITDIR` and then `git -C $gitdir config --get annex.uuid` is an OK thing to do or an ugly hack....

yarikoptic commented:
why not just some really temporary name but near the target, e.g. `$(mktemp "${EXPORT_LOC}.tmp-XXXXXXXX")`. This way in principle you could later identify some junk left around if interrupted etc. BTW -- what if interrupted etc - try to `rclone delete` also after `TRANSFER-FAILURE STORE "$key" "upload failed"`

Wolf480pl commented:
AFAIU mktemp works on local filesystems because it uses `O_EXCL`. Using it to generate a random name for rclone would rely on randomness, which can be not-very-random on sth like a RasPi.... I can do it that way if you think it's good enough...

yarikoptic commented:
I think it would be random enough unless someone shows pragmatically otherwise, eg demonstrate a collision in real world case. There is also `$RANDOM` in bash IIRC, might suffice here

Wolf480pl committed:
tests: Add tests of export feature

Wolf480pl commented:
I guess I'll use mktemp, should be random enough.

Also, while trying to fix that I hit what seems like a bug in git annex https://git-annex.branchable.com/bugs/external_remote_export_sent_to_wrong_process/

Wolf480pl committed:
export: use /dev/urandom for tmploc

Key isn't unique enough, since we can be uploading multiple copies
of the same file with different names, and collision of identical files
*is* an issue because we *move* the files when we're done with them,
so the process that loses the race can't move the file it uploaded.

We could use key + local repo UUID + hash(export_loc) to get sth unique,
but getting those is kinda hacky.

Let's use a hash of 128 bits from /dev/urandom instead.

Since we're using randomness anyway, basing tmploc on export_loc
doesn't hurt, and also makes it more likely that tmploc and export_loc
will be on the same filesystem.

Wolf480pl commented:
Went with `/dev/urandom` instead.

Unfortunately, the tests are now using git-annex version 10.20230329 despite specifying `:release`, so I can't test the safeguards I against the git-annex race condition that I was planning to add.

Wolf480pl committed:
export: defend against EXPORT being send to the wrong process

Check that EXPORT is sent exactly once
before each {TRANSFER,CHECK,REMOVE,RENAME}EXPORT
and scream loudly otherwise.

yarikoptic commented:
> Went with `/dev/urandom` instead.

ok, although I am still not sure why you really need good "source of randomness" here -- we are talking a random suffix to the target file, how can you really get a collision there since even parallel git-annex worker processes in some `annex -J` invocation unlikely to be given the same key/file for processing.  Did you really observe some situation/collision due to weak randomness there or what am I missing?

Arguments against using it which I had to dismiss myself
- `/dev/urandom` makes it a bit too OS specific to my liking.  I would not be surprised if someone at some point uses on some exotic system with bash.  Well, https://en.wikipedia.org/wiki//dev/random suggests it being universally present, even under cygwin.  Not yet sure if under msys2 but not sure if this `git-annex-remote` can be used there anyways. We might need to setup test env. chatgpt says that it is available in msys2.  But also it says "It's possible for an environment to have Bash available but not have the /dev/urandom device file. For example, some embedded systems or minimalist Linux distributions may not include /dev/urandom by default."  So, overall, pragmatically speaking it should be present.
- It is blocking until enough randomness.  On my test on laptop it was taking 1-2 ms to get those 16 bytes. Not sure if there would be a system which might be considerably slower thus leading to performance impact.

Submitting #66 which, if finished, should give us the msys2 test env to guarantee compatibility there.

> Unfortunately, the tests are now using git-annex version 10.20230329 despite specifying `:release`, so I can't test the safeguards I against the git-annex race condition that I was planning to add.

well, there was a `:release` yesterday (https://github.com/datalad/git-annex/releases/tag/10.20230329) so it does as you say ;)  Do you need to test in some older one?  then we might need to add a dedicated single env with some older annex of our choice.  I think it would be worthwhile anyways so we could guarantee that we stay compatible with .  Submitting #65

Wolf480pl commented:
>  Do you need to test in some older one?

I managed to figure out with great pain how to install an older version of git-annex in the macos CI run, and was able to test the bug safeguard https://github.com/Wolf480pl/git-annex-remote-rclone/actions/runs/4568849535

> ok, although I am still not sure why you really need good "source of randomness" here

Because I was convinced that if it collides, there could be data loss.

Consider the following:

P1: starts uploading foo.tmp-1234
P2: starts uploading foo.tmp-1234 from a different commit (it's in a different repo)
P1: finishes uploading, moves foo.tmp-1234 to foo
P2: still uploading
P3: runs CHECKPRESENTEXPORT on foo, returns success 

This violates [the protocol](https://git-annex.branchable.com/design/external_special_remote_protocol/export_and_import_appendix/) which says

> Note that it's important that, while a file is being stored, CHECKPRESENTEXPORT not indicate it's present until all the data has been transferred.

Now that I think of it, I can't prove that this can lead to data loss, but I can't prove that it can't either.

Am I taking the "people are entrusting us with their data" too seriously?

Wolf480pl commented:
Though now that I think of it: if P1 and P2 are writing the same content, then the file that does get stored should be fine, and the other file will have the move fail, and git-annex will know that it's not stored...

Maybe sha1(key + $RANDOM) ?

yarikoptic commented:
> > Do you need to test in some older one?
> 
> I managed to figure out with great pain how to install an older version of git-annex in the macos CI run, and was able to test the bug safeguard https://github.com/Wolf480pl/git-annex-remote-rclone/actions/runs/4568849535

in master we now test against older one on linux, I didn't try to install one via brew but in principle should be as easy via `datalad/git-annex:release` approach and just specify the version.  In above run I see indeed quite elaborate

```
cd $(brew --repo) && ls Library && mkdir -p Library/Formula && wget -O Library/Formula/git-annex.rb https://raw.githubusercontent.com/Homebrew/homebrew-core/6a232aa8b4525cdc102152a6c646572eb09863ed/Formula/git-annex.rb && brew install git-annex
```
which installs `git-annex--10.20230227.monterey.bottle.tar.gz` .

> Maybe `${EXPORT_LOC}.tmp-$(echo "$RANDOM $key" |git hash-object)` ?

why not simply `${EXPORT_LOC}.tmp-$RANDOM`  or `${EXPORT_LOC}.tmp-$RANDOM$RANDOM`  for "extra protection" ;-) ?

Wolf480pl commented:
> Though now that I think of it: if P1 and P2 are writing the same content, then the file that does get stored should be fine

Nvm it won't, not if the remote end truncates the file when opening for writing.

> why not simply ${EXPORT_LOC}.tmp-$RANDOM or ${EXPORT_LOC}.tmp-$RANDOM$RANDOM for "extra protection" ;-) ?

In the scenario where two different people are uploading files to the same location at the same time, what is the acceptable probability of data loss?

yarikoptic commented on git-annex-remote-rclone:
```diff
@@ -258,18 +285,102 @@ while read -r line; do
+					else
+						if ! runcmd rclone moveto "$tmploc" "$EXPORT_LOC"; then
+							runcmd rclone delete "$tmploc"
+							echo TRANSFER-FAILURE STORE "$key" "move to final destination failed"
```
BTW -- shouldn't we in any case of failure try to remove that `$tmploc` to clean up after ourselves?

yarikoptic replied:
@Wolf480pl ping on this

Wolf480pl replied:
Sorry, I didn't have time recently, I'll try to look into it later today

yarikoptic replied:
sorry to nudge, but just today while initiating a new remote for a backup of singularity hub I thought -- wouldn't it be nice if we could export to dropbox... which reminded me about this PR ;)  I really hope you would find time to bring this PR over the finish line, especially since IIRC it is so close.

yarikoptic commented:
I thought you figured it should not happen anyways due to:

> Though now that I think of it: if P1 and P2 are writing the same content, then the file that does get stored should be fine, and the other file will have the move fail, and git-annex will know that it's not stored..

re:

> > why not simply ${EXPORT_LOC}.tmp-$RANDOM or ${EXPORT_LOC}.tmp-$RANDOM$RANDOM for "extra protection" ;-) ?
> 
> In the scenario where two different people are uploading files to the same location at the same time, what is the acceptable probability of data loss in the form of the uploaded file being corrupted while git-annex thinks it's correctly uploaded?

I think that if two people at the same time exporting into the same location we might have bigger problems if there is no protection at git-annex level for that (e.g. some sentinel "lock" file to inform that export is ongoing). I didn't check. But I just think that striving so hard to avoid collision here is more of red-herring but sure thing I can be wrong, thus if you really feel that more complex `${EXPORT_LOC}.tmp-$(echo "$RANDOM $key" |git hash-object)` would help and should not hurt (e.g. making it too long for that remote filesystem), we can try going that way. But typically I think "simpler/more transparent" is better/easier to debug.

yarikoptic commented:
ok -- let's say you win! -- `$RANDOM` is indeed not really random enough probably

<details>
<summary>Running this simple script where parallel processes generate some number of `$RANDOMs`</summary> 

```shell
#!/bin/bash
set -eu

cd "$(mktemp -d ${TMPDIR:-/tmp}/dl-XXXXXXX)"

echo messing with randoms under $PWD
echo $RANDOM > random

function random_loop_check() {
    for l in {1..10000}; do
        r=$RANDOM
        if grep "^$r" random; then
            echo "COLLISION!!"
            exit 1
        fi
        echo $r >| random.$r
        mv random.$r random
    done
    echo done here with last random $r
}


for i in {1..10}; do
    echo "starting $i one"
    random_loop_check &
done

wait

```
</details>

<details>
<summary>produces this output with at least 5 collisions</summary> 

```shell
❯ bash collide-random.sh
messing with randoms under /home/yoh/.tmp/dl-UzG96kN
starting 1 one
starting 2 one
starting 3 one
starting 4 one
starting 5 one
starting 6 one
starting 7 one
starting 8 one
starting 9 one
starting 10 one
27014
COLLISION!!
mv: cannot stat 'random.20814': No such file or directory
mv: cannot stat 'random.29022': No such file or directory
1443
COLLISION!!
mv: cannot stat 'random.1014': No such file or directory
8936
COLLISION!!
20224
COLLISION!!
15381
COLLISION!!
done here with last random 23327
done here with last random 10315
bash collide-random.sh  51.65s user 20.39s system 511% cpu 14.085 total

```
</details>

with `$RANDOM$RANDOM` there I get no collisions in this "10 people in parallel at exactly the same times very quickly uploading 10000 files ;)" scenario ;)

yarikoptic commented:
ping @Wolf480pl let's get this PR finished?

cnjr2 commented:
I would love for this to be available.

My scenario is to have a media server that accesses data via an rclone mount of an S3 remote. I want to populate the remote via git annex with a file tree so that the media server will show files with names that make sense. This would all be possible with a git-annex special remote of type S3 and I could cut out rclone all together, but what if I want my files to be chunked and encrypted? I only see that work with rclone handling all the data transfer.
